### PR TITLE
#164980291-Ft-implement user signup endpoint

### PR DESCRIPTION
**Pull request by Joyce-O**
# Description
This PR creates user signup functionality. It does not generate a token yet.

# Type of change
feature
- add user model
- add user route
- add user controller

# Pivotal Tracker ID
#164980291

### CODE REVIEW

**Comment by Jesse-efe** [comment]
You could squash the three commits into one so that the PR contains just one commit

File: src/controller/auth.js [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+    req.body = JSON.parse(JSON.stringify(req.body));
+    return user
+      .create({
+        username: req.body.userName,
+        firstname: req.body.firstName,
+        lastname: req.body.lastName,
+        email: req.body.email,
+        bio: req.body.hasOwnProperty('bio') ? req.body.bio : null,
+        image: req.body.hasOwnProperty('image') ? req.body.image : null,
+        password: req.body.password,
```
Comment (fuchodeveloper) [inline src/controller/auth.js]: You could hash the password before saving to the database. That way, it is not stored as plain text

File: src/controller/auth.js [hunk 2]
```diff
@@ -0,0 +1,19 @@
+const user = require('../models').user;
+
+module.exports = {
+  create(req, res) {
+    req.body = JSON.parse(JSON.stringify(req.body));
```
Comment (fuchodeveloper) [inline src/controller/auth.js]: You could incorporate validation before processing the user input

**Review by fuchodeveloper — CHANGES_REQUESTED** [review]
Resolve HoundCI violations, merge conflicts and take a look at the suggested changes.

File: src/controller/auth.js (hunk 2 above)
Comment (Joyce-O) [reply src/controller/auth.js]: Okay, I will consume the validation module one of the team members have created already.

File: src/controller/auth.js (hunk 1 above)
Comment (Joyce-O) [reply src/controller/auth.js]: Yes, I will do that now

**Review by Fiyiin — COMMENTED** [review]
Hey Joyce, check your commits, it seems you have 3

File: src/controller/auth.js (hunk 2 above)
Comment (profmcdan) [reply src/controller/auth.js]: This comment has been here for four days, I was wondering why you have not resolved it.

File: src/controller/auth.js (hunk 1 above)
Comment (profmcdan) [reply src/controller/auth.js]: Also pending

File: src/migration/20190328072153-create-user.js [hunk 3]
```diff
@@ -0,0 +1,50 @@
...
+      },
+      firstName: {
+        type: Sequelize.STRING(255),
+        field: 'first_name',
+      },
+      lastName: {
+        type: Sequelize.STRING(255),
+        field: 'last_name',
+      },
+      email: {
```
[the remark this answers is not in the document]
Comment (profmcdan) [reply src/migration/20190328072153-create-user.js]: Kindly attend to houndci review almost immediately

File: src/routes/api/users/auth.js [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    } catch (err) {
+      return response.status(409).json({
+        status: 409,
+        error: err.message,
+      });
+    }
+  }
+
+  static async 
+}
```
[the remark this answers is not in the document]
Comment (Joyce-O) [reply src/routes/api/users/auth.js:56]: Resolved

**Comment by Bobjayafam** [comment]
LGTM

File: src/routes/api/users/users_route.js [hunk 5]
```diff
@@ -0,0 +1,10 @@
+import express from 'express';
+import validate from '../../../middlewares/validator';
+import auth from './auth';
+
+
+const users = express.Router();
+
+users.post('/users', validate.userSignup, auth);
+
+export default users;
```
Comment (fuchodeveloper) [inline src/routes/api/users/users_route.js]: The default export should match the filename. Also, use camel case for filenames.

File: test/test.js [hunk 6]
```diff
@@ -149,3 +154,24 @@ describe('Tests for password hashing functionality', () => {
...
+      nock('http://testwiththisurl.com')
+        .get('/users')
+        .reply(201, user);
+    });
+    it('should return the expected json response', async () => {
+      const response = await supertest(app)
+        .get('/users');
+      assert(response.statusCode, 201);
+      expect(user).to.be.an('object');
+      expect(undefined).to.be.an('undefined');
```
Comment (fuchodeveloper) [inline test/test.js]: Add a test for success and failure on registration. You could test for the actually returned values in the object, that way your tests are more concrete.

**Review by fuchodeveloper — CHANGES_REQUESTED** [review]

File: src/routes/api/users/users_route.js (hunk 5 above)
Comment (Joyce-O) [reply src/routes/api/users/users_route.js]: Done

**Comment by Fiyiin** [comment]
LGTM

**Comment by kossy360** [comment]
LGTM

**Comment by Bobjayafam** [comment]
LGTM

**Comment by Jesse-efe** [comment]
LGTM

**Comment by Joyce-O** [comment]
All the identified errors has been resolved

**Pull request merged**